### PR TITLE
Backport upstream changes

### PR DESCRIPTION
**Pull request by ctavan**

### CODE REVIEW

File: src/util/printValue.js [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import isFunction from 'lodash/isFunction';
 import isSymbol from 'lodash/isSymbol';
 
-const symbolToString = Symbol.prototype.toString;
+const symbolToString = typeof Symbol !== 'undefined' ?
```
Comment (mborst) [inline src/util/printValue.js]: Is this for backwards compatibility? Does `typeof` really return `undefined` if the global `Symptom` does not exist? Shouldn't `Symbol` be a function or constructor to have `Symbol.prototype.toString`? I don't understand.

File: src/string.js [hunk 2]
```diff
@@ -84,19 +86,21 @@ inherits(StringSchema, MixedSchema, {
   },
 
   matches(regex, options = {}) {
-    let excludeEmptyString = false;
-    let message;
+    const optionsObject = isString(options) ? {} : options;
```
Comment (mborst) [inline src/string.js]: What this flubber interface does to the parameter handling...

File: src/string.js [hunk 3]
```diff
@@ -84,19 +86,21 @@ inherits(StringSchema, MixedSchema, {
...
-    } else { message = options; }
+    const { excludeEmptyString, message } = { ...defaultOptions, ...optionsObject };
 
     return this.test({
       message: message || locale.matches,
       params: { regex },
       test: value => (
         isAbsent(value) ||
-        (value === '' && excludeEmptyString) ||
+        (value === '' && !excludeEmptyString) ||
```
Comment (mborst) [inline src/string.js]: Maybe separate bugfix and refactor?

File: test/string.js [hunk 4]
```diff
@@ -112,19 +112,43 @@ describe('String types', () => {
...
   });
 
   it('MATCHES should exclude empty strings', () => {
     const v = string().matches(/(hi|bye)/, { excludeEmptyString: true });
 
-    return v.isValid('').should.eventually().equal(true);
+    return v.isValid('').should.eventually().equal(false);
+  });
+
+  it('MATCHES should populate the error message', async () => {
```
Comment (mborst) [inline test/string.js]: Maybe add these tests before you do the changes?

File: src/util/printValue.js (hunk 1 above)
Comment (ctavan) [reply src/util/printValue.js]: I've added the related issue jquense/yup#131 in the commit message.

**Pull request merged**